quota/quota.go

```diff
@@ -0,0 +1,180 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package quota
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+)
+
+// Group is a quota group of snaps, services or sub-groups that are all subject
+// to specific resource quotas. Current quota resource types supported are just
+// memory, but this can be expanded in the future.
+type Group struct {
+	// Name is the name of the quota group. Certain names are reserved and have
+	// special meaning such as user.* or system.* or snapd.*, but are otherwise
+	// generically choosable by a user. This name will correspond to the name of
+	// the systemd slice underlying the quota group.
+	Name string
+
+	// SubGroups is the set of sub-groups that are subject to this quota.
+	// Sub-groups have their own limits, subject to the requirement that the
+	// highest quota for a sub-group is that of the parent group.
+	SubGroups []*Group `json:"sub-groups,omitempty"`
+
+	// MemoryLimit is the limit of memory available to the processes in the
+	// group where if the total used memory of all the processes exceeds the
+	// limit, oom-killer is invoked which will start killing processes. The
+	// specific behavior of which processes are killed is subject to the
+	// ExhaustionBehavior.
+	MemoryLimit quantity.Size `json:"memory-limit,omitempty"`
+
+	// MemoryExhaustionBehavior is the behavior that is implemented to decide
+	// what processes to kill in the quota group when the memory limits are
+	// exhausted.
+	MemoryExhaustionBehavior MemoryExhaustionBehavior `json:"memory-exhaustion-behavior,omitempty"`
+
+	// parentGroup is the the parent group that this group is a child of. It is
+	// here mainly for coding convenience to easily construct the full chain
+	// back to a root group to generate the full slice unit name of the quota
+	// group. If it is nil, then this is a "root" quota group.
+	parentGroup *Group
+}
+
+// MemoryExhaustionBehavior is the behavior determining what the system should
+// do when memory available to a quota group is exhausted.
+type MemoryExhaustionBehavior int
+
+const (
+	// DefaultOOMKiller is the default behavior to invoke to choose which
+	// processes when the memory quota resource limit is reached, and it
+	// consists of letting the Linux kernel's oom-killer run and decide, which
+	// is available on all cgroups version systems, but is not deterministic as
+	// to which processes are killed.
+	DefaultOOMKiller MemoryExhaustionBehavior = iota
+)
+
+// SliceFileName returns the name of the slice file that should be used for this
+// quota group. This name will include all of the group's parents in the name.
+// For example, a group named "bar" that is a child of the "foo" group will have
+// a systemd slice name as "foo-bar.slice". Note that the slice name may differ
```

Comment:
Should we include `snap.` prefix in the slice names?

Reply:
Yes, we now include snap. prefix in slice names